server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -2034,6 +2037,8 @@ static CanMatchShardResponse canMatch(CanMatchContext canMatchContext, boolean c
                     indexService = readerContext.indexService();
                     QueryRewriteContext queryRewriteContext = canMatchContext.getQueryRewriteContext(indexService);
                     if (queryStillMatchesAfterRewrite(canMatchContext.request, queryRewriteContext) == false) {
+                        indexService.getSearchOperationListener()
```

Comment:
This particular construct repeats in several places, maybe make a private method out of it? Or just place it in the `finally` clause? Or maybe both?